File: apps/java/src/main/java/com/inthebeginning/simulator/Environment.java
package com.inthebeginning.simulator;

import java.util.*;

import static com.inthebeginning.simulator.Constants.*;

/**
 * Environmental system modeling planetary conditions.
 * Tracks temperature, atmosphere composition, radiation levels,
 * ocean chemistry, and geological activity that influence chemical
 * and biological evolution.
 */
public class Environment {

    /** Atmospheric composition as gas name -> fraction. */
    private final Map<String, Double> atmosphere = new LinkedHashMap<>();

    /** Ocean properties. */
    private double oceanCoverage = 0.0;   // fraction 0-1
    private double oceanPH = 7.0;
    private double oceanTemperature = 0.0;

    /** Surface conditions. */
    private double surfaceTemperature;
    private double pressure = 1.0;         // atmospheres
    private double uvIntensity = 0.0;
    private double cosmicRayFlux = 0.0;
    private double volcanicActivity = 0.0;
    private double magneticFieldStrength = 0.0;

    /** Energy sources. */
    private double solarLuminosity = 0.0;
    private double geothermalEnergy = 0.0;
    private double lightningFrequency = 0.0;

    /** Habitability score 0-1. */
    private double habitability = 0.0;

    private final Random rng;

    public Environment(Random rng) {
        this.surfaceTemperature = 0.0;
        this.rng = rng;
    }

    // --- Accessors ---

    public Map<String, Double> getAtmosphere()     { return atmosphere; }
    public double getOceanCoverage()               { return oceanCoverage; }
    public double getOceanPH()                     { return oceanPH; }
    public double getOceanTemperature()            { return oceanTemperature; }
    public double getSurfaceTemperature()          { return surfaceTemperature; }
    public double getPressure()                    { return pressure; }
    public double getUvIntensity()                 { return uvIntensity; }
    public double getCosmicRayFlux()               { return cosmicRayFlux; }
    public double getVolcanicActivity()            { return volcanicActivity; }
    public double getMagneticFieldStrength()        { return magneticFieldStrength; }
    public double getSolarLuminosity()             { return solarLuminosity; }
    public double getGeothermalEnergy()            { return geothermalEnergy; }
    public double getLightningFrequency()          { return lightningFrequency; }
    public double getHabitability()                { return habitability; }

    public void setSurfaceTemperature(double t)    { this.surfaceTemperature = t; }
    public void setPressure(double p)              { this.pressure = p; }

    // --- Planetary formation ---

    /**
     * Initialize early Earth conditions:
     * reducing atmosphere, high volcanic activity, no oceans yet.
     */
    public void initializeEarlyEarth() {
        surfaceTemperature = 500.0; // hot early Earth
        pressure = 10.0;           // thick early atmosphere

        atmosphere.clear();
        atmosphere.put("CO2", 0.60);
        atmosphere.put("N2", 0.20);
        atmosphere.put("H2O", 0.10);
        atmosphere.put("CH4", 0.05);
        atmosphere.put("NH3", 0.03);
        atmosphere.put("H2", 0.02);

        volcanicActivity = 0.9;
        geothermalEnergy = 100.0;
        magneticFieldStrength = 0.5;
        uvIntensity = 0.8;        // no ozone layer yet
        cosmicRayFlux = 0.3;
        solarLuminosity = 0.7;    // young sun was dimmer
        lightningFrequency = 0.8;

        oceanCoverage = 0.0;
        habitability = 0.0;
    }

    /**
     * Cool the planet and form oceans.
     * Called over multiple ticks to simulate geological cooling.
     */
    public void geologicalCooling(double coolingRate) {
        if (surfaceTemperature > T_EARTH_SURFACE) {
            surfaceTemperature -= coolingRate;
            if (surfaceTemperature < T_EARTH_SURFACE) {
                surfaceTemperature = T_EARTH_SURFACE;
            }
        }

        // Oceans form when temperature drops below 373K (boiling point)
        if (surfaceTemperature < 373.0 && oceanCoverage < 0.71) {
            oceanCoverage += 0.01;
            if (oceanCoverage > 0.71) oceanCoverage = 0.71;
            oceanTemperature = surfaceTemperature - 10;
            oceanPH = 5.5 + rng.nextGaussian() * 0.2; // early oceans were acidic
        }

        // Volcanic activity decreases over time
        if (volcanicActivity > 0.1) {
            volcanicActivity *= 0.999;
        }

        // Pressure normalizes
        if (pressure > 1.0) {
            pressure *= 0.999;
            if (pressure < 1.0) pressure = 1.0;
        }

        updateHabitability();
    }

    /**
     * Atmosphere evolves: CO2 dissolves in oceans, O2 builds up from photosynthesis.
     */
    public void evolveAtmosphere(boolean photosynthesisActive) {
        // CO2 dissolves in oceans
        if (oceanCoverage > 0.1) {
            double co2 = atmosphere.getOrDefault("CO2", 0.0);
            if (co2 > 0.001) {
                atmosphere.put("CO2", co2 * 0.999);
            }
        }

        // Photosynthesis produces O2
        if (photosynthesisActive) {
            double o2 = atmosphere.getOrDefault("O2", 0.0);
            atmosphere.put("O2", Math.min(0.21, o2 + 0.0001));

            // O2 -> ozone reduces UV
            if (o2 > 0.01) {
                uvIntensity = Math.max(0.05, uvIntensity * 0.999);
            }
        }

        // Normalize atmosphere fractions
        double total = atmosphere.values().stream().mapToDouble(Double::doubleValue).sum();
        if (total > 0) {
            for (Map.Entry<String, Double> entry : atmosphere.entrySet()) {
                entry.setValue(entry.getValue() / total);
            }
        }

        updateHabitability();
    }

    /**
     * Compute a composite habitability score.
     * Considers temperature, water, atmosphere, and radiation.
     */
    private void updateHabitability() {
        double tempScore = 0.0;
        if (surfaceTemperature > 273 && surfaceTemperature < 373) {
            // Optimal around 288K
            tempScore = 1.0 - Math.abs(surfaceTemperature - 288.0) / 85.0;
            tempScore = Math.max(0, Math.min(1, tempScore));
        }

        double waterScore = Math.min(1.0, oceanCoverage / 0.5);

        double atmosphereScore = 0.0;
        double n2 = atmosphere.getOrDefault("N2", 0.0);
        double o2 = atmosphere.getOrDefault("O2", 0.0);
        double co2 = atmosphere.getOrDefault("CO2", 0.0);
        atmosphereScore = Math.min(1.0, n2 + o2 * 2.0);
        if (co2 > 0.5) atmosphereScore *= 0.5; // too much CO2 is bad

        double radiationScore = 1.0 - Math.min(1.0, uvIntensity);

        habitability = (tempScore * 0.3 + waterScore * 0.3 +
                        atmosphereScore * 0.2 + radiationScore * 0.2);
    }

    /**
     * Determine environmental pressure on organisms.
     * Higher = harder to survive.
     */
    public double environmentalPressure() {
        double pressure = 0.0;
        pressure += uvIntensity * UV_MUTATION_RATE * 1000;
        pressure += cosmicRayFlux * COSMIC_RAY_MUTATION_RATE * 1000;
        pressure += Math.max(0, (surfaceTemperature - 373) / 100.0);
        pressure += Math.max(0, (273 - surfaceTemperature) / 100.0);
        pressure += volcanicActivity * 0.1;
        return Math.min(0.9, pressure);
    }

    /**
     * Check whether conditions support liquid water.
     */
    public boolean hasLiquidWater() {
        return surfaceTemperature > 273 && surfaceTemperature < 373
                && oceanCoverage > 0.01 && pressure > 0.006;
    }

    /**
     * Check whether conditions can support complex life.
     */
    public boolean supportsComplexLife() {
        return habitability > 0.5
                && hasLiquidWater()
                && atmosphere.getOrDefault("O2", 0.0) > 0.05
                && uvIntensity < 0.3;
    }

    public String toCompact() {
        StringBuilder atmo = new StringBuilder();
        atmosphere.entrySet().stream()
                .sorted((a, b) -> Double.compare(b.getValue(), a.getValue()))
                .limit(4)
                .forEach(e -> {
                    if (atmo.length() > 0) atmo.append(",");
                    atmo.append(String.format("%s:%.1f%%", e.getKey(), e.getValue() * 100));
                });

        return String.format(
                "ENV[T=%.0fK P=%.1fatm ocean=%.0f%% UV=%.2f hab=%.2f atmo={%s}]",
                surfaceTemperature, pressure, oceanCoverage * 100,
                uvIntensity, habitability, atmo);
    }
}

File: apps/kotlin/app/src/main/java/com/inthebeginning/simulator/Environment.kt
package com.inthebeginning.simulator

import com.inthebeginning.simulator.Constants.COSMIC_RAY_MUTATION_RATE
import com.inthebeginning.simulator.Constants.K_B
import com.inthebeginning.simulator.Constants.RADIATION_DAMAGE_THRESHOLD
import com.inthebeginning.simulator.Constants.T_CMB
import com.inthebeginning.simulator.Constants.T_EARTH_SURFACE
import com.inthebeginning.simulator.Constants.T_PLANCK
import com.inthebeginning.simulator.Constants.THERMAL_FLUCTUATION
import com.inthebeginning.simulator.Constants.UV_MUTATION_RATE
import kotlin.math.exp
import kotlin.random.Random

/**
 * Atmospheric composition as gas fractions.
 */
data class Atmosphere(
    var nitrogen: Double = 0.0,
    var oxygen: Double = 0.0,
    var carbonDioxide: Double = 0.0,
    var methane: Double = 0.0,
    var water: Double = 0.0,
    var hydrogen: Double = 0.0,
    var helium: Double = 0.0,
    var ammonia: Double = 0.0
) {
    /** Total atmospheric pressure (arbitrary units). */
    val totalPressure: Double
        get() = nitrogen + oxygen + carbonDioxide + methane +
                water + hydrogen + helium + ammonia

    /** Whether this atmosphere is reducing (no free oxygen). */
    val isReducing: Boolean get() = oxygen < 0.01 && (hydrogen + methane + ammonia) > 0.1

    /** Whether this atmosphere is oxidizing (has free oxygen). */
    val isOxidizing: Boolean get() = oxygen > 0.01

    /** Greenhouse effect factor based on greenhouse gases. */
    val greenhouseEffect: Double
        get() = 1.0 + carbonDioxide * 30.0 + methane * 80.0 + water * 5.0
}

/**
 * Represents a body of liquid on the planetary surface.
 */
data class OceanState(
    var volume: Double = 0.0,
    var salinity: Double = 0.0,
    var pH: Double = 7.0,
    var dissolvedMinerals: Double = 0.0
) {
    val isPresent: Boolean get() = volume > 0.01
}

/**
 * Radiation type impinging on the environment.
 */
enum class RadiationType(val displayName: String) {
    STELLAR_UV("Stellar UV"),
    COSMIC_RAYS("Cosmic Rays"),
    RADIOACTIVE_DECAY("Radioactive Decay"),
    THERMAL_IR("Thermal IR");
}

/**
 * Represents a single radiation source.
 */
data class RadiationSource(
    val type: RadiationType,
    var intensity: Double,
    val mutagenicFactor: Double
)

/**
 * Environmental conditions for the simulation.
 * Models temperature, radiation, atmosphere, and geological activity
 * as they evolve through cosmic epochs.
 */
class Environment(
    var temperature: Double = T_PLANCK
) {
    var radiation: Double = 0.0
    var cosmicAge: Double = 0.0
    var magneticFieldStrength: Double = 0.0
    var tectonicActivity: Double = 0.0

    val atmosphere: Atmosphere = Atmosphere()
    val ocean: OceanState = OceanState()

    val radiationSources: MutableList<RadiationSource> = mutableListOf()
    val events: MutableList<String> = mutableListOf()

    private val rng = Random

    /** Whether the environment has liquid water. */
    val hasLiquidWater: Boolean
        get() = temperature in 273.0..373.0 && ocean.isPresent

    /** Whether lightning is occurring (requires atmosphere and thermal gradient). */
    val hasLightning: Boolean
        get() = atmosphere.totalPressure > 0.5 &&
                temperature in 250.0..400.0 &&
                rng.nextDouble() < 0.1

    /** UV radiation level at the surface. */
    val uvLevel: Double
        get() {
            val stellarUV = radiationSources
                .filter { it.type == RadiationType.STELLAR_UV }
                .sumOf { it.intensity }
            // Ozone layer reduces UV (oxygen in atmosphere)
            val ozoneShielding = 1.0 - (atmosphere.oxygen * 5.0).coerceIn(0.0, 0.95)
            return stellarUV * ozoneShielding
        }

    /** Cosmic ray flux at the surface. */
    val cosmicRayFlux: Double
        get() {
            val crSources = radiationSources
                .filter { it.type == RadiationType.COSMIC_RAYS }
                .sumOf { it.intensity }
            // Magnetic field deflects cosmic rays
            val shielding = 1.0 - (magneticFieldStrength * 0.8).coerceIn(0.0, 0.95)
            return crSources * shielding
        }

    /** Effective mutation rate from all radiation sources. */
    val mutationRate: Double
        get() = uvLevel * UV_MUTATION_RATE + cosmicRayFlux * COSMIC_RAY_MUTATION_RATE

    /**
     * Initialize as the early universe (pre-stellar).
     */
    fun initEarlyUniverse() {
        temperature = T_PLANCK
        radiation = T_PLANCK * K_B
        atmosphere.hydrogen = 0.75
        atmosphere.helium = 0.25
        magneticFieldStrength = 0.0
        tectonicActivity = 0.0
        events.add("Environment initialized: early universe")
    }

    /**
     * Set conditions for early Earth (~4.5 Ga).
     */
    fun initEarlyEarth() {
        temperature = 500.0 // Hot, volcanic
        radiation = 5.0
        magneticFieldStrength = 0.5
        tectonicActivity = 0.8

        atmosphere.nitrogen = 0.0
        atmosphere.oxygen = 0.0
        atmosphere.carbonDioxide = 0.7
        atmosphere.methane = 0.1
        atmosphere.water = 0.1
        atmosphere.hydrogen = 0.05
        atmosphere.ammonia = 0.05

        radiationSources.clear()
        radiationSources.add(RadiationSource(RadiationType.STELLAR_UV, 3.0, UV_MUTATION_RATE))
        radiationSources.add(RadiationSource(RadiationType.COSMIC_RAYS, 1.0, COSMIC_RAY_MUTATION_RATE))
        radiationSources.add(RadiationSource(RadiationType.RADIOACTIVE_DECAY, 2.0, UV_MUTATION_RATE * 0.5))

        events.add("Early Earth conditions set: reducing atmosphere")
    }

    /**
     * Cool Earth and form oceans.
     */
    fun formOceans() {
        if (temperature > 373.0) return

        ocean.volume = atmosphere.water * 100.0
        atmosphere.water *= 0.1 // Most water condenses
        ocean.pH = 5.5 // Slightly acidic early ocean
        ocean.salinity = 0.1
        ocean.dissolvedMinerals = 0.5

        events.add("Oceans formed: liquid water present")
    }

    /**
     * Great Oxidation Event: oxygen from photosynthesis accumulates.
     */
    fun greatOxidationEvent(oxygenProduction: Double) {
        val freeOxygen = oxygenProduction * 0.01
        atmosphere.oxygen += freeOxygen

        // Methane reacts with oxygen
        val methaneOxidized = (atmosphere.methane * atmosphere.oxygen * 0.1)
            .coerceAtMost(atmosphere.methane)
        atmosphere.methane -= methaneOxidized
        atmosphere.carbonDioxide += methaneOxidized * 0.5

        if (atmosphere.oxygen > 0.01) {
            events.add("Great Oxidation Event: O2 = ${String.format("%.4f", atmosphere.oxygen)}")
        }
    }

    /**
     * Evolve environmental conditions by one time step.
     */
    fun evolve(tick: Int) {
        cosmicAge = tick.toDouble()

        // Universal cooling
        when {
            tick < Constants.RECOMBINATION_EPOCH -> {
                temperature = T_PLANCK * exp(-tick.toDouble() / 10000.0)
            }
            tick < Constants.STAR_FORMATION_EPOCH -> {
                temperature = (T_CMB + (3000.0 - T_CMB) *
                        exp(-(tick - Constants.RECOMBINATION_EPOCH).toDouble() / 50000.0))
            }
            tick < Constants.EARTH_EPOCH -> {
                // Background stays at CMB temperature
                temperature = T_CMB
            }
            tick < Constants.LIFE_EPOCH -> {
                // Earth surface temperature
                val cooling = exp(-(tick - Constants.EARTH_EPOCH).toDouble() / 20000.0)
                temperature = T_EARTH_SURFACE + 200.0 * cooling
                if (temperature < 373.0 && !ocean.isPresent) {
                    formOceans()
                }
            }
            else -> {
                // Stabilize near modern Earth temperature
                temperature = T_EARTH_SURFACE + rng.nextGaussian() * THERMAL_FLUCTUATION * 10.0
            }
        }

        // Thermal fluctuation
        temperature += rng.nextGaussian() * THERMAL_FLUCTUATION

        // Radiation decay
        for (source in radiationSources) {
            source.intensity *= 0.99999
        }

        // Update radiation total
        radiation = radiationSources.sumOf { it.intensity }

        // Tectonic activity slowly decreases
        tectonicActivity *= 0.99999

        // Nitrogen slowly builds up from volcanic outgassing
        if (tick > Constants.EARTH_EPOCH && atmosphere.nitrogen < 0.78) {
            atmosphere.nitrogen += tectonicActivity * 0.00001
        }
    }

    /**
     * Check if conditions are suitable for life.
     */
    val isHabitable: Boolean
        get() = hasLiquidWater &&
                temperature in 250.0..400.0 &&
                radiation < RADIATION_DAMAGE_THRESHOLD

    /**
     * Environmental stress factor (0 = ideal, higher = more stressful).
     */
    val stressFactor: Double
        get() {
            var stress = 0.0
            if (temperature !in 273.0..330.0) {
                stress += kotlin.math.abs(temperature - 300.0) / 100.0
            }
            stress += radiation * 0.1
            stress += uvLevel * 0.5
            return stress.coerceAtLeast(0.0)
        }

    /**
     * Apply a random catastrophic event with low probability.
     */
    fun randomCatastrophe(): String? {
        val roll = rng.nextDouble()
        return when {
            roll < 0.0001 -> {
                // Asteroid impact
                temperature += 50.0
                radiation += 5.0
                events.add("Asteroid impact!")
                "Asteroid impact"
            }
            roll < 0.0003 -> {
                // Supervolcano
                atmosphere.carbonDioxide += 0.05
                temperature += 10.0
                events.add("Supervolcano eruption!")
                "Supervolcano eruption"
            }
            roll < 0.0005 -> {
                // Solar flare
                radiation += 3.0
                events.add("Massive solar flare!")
                "Massive solar flare"
            }
            else -> null
        }
    }

    /** Summary of current environmental conditions. */
    fun summary(): Map<String, Any> = mapOf(
        "temperature" to temperature,
        "radiation" to radiation,
        "hasLiquidWater" to hasLiquidWater,
        "isHabitable" to isHabitable,
        "atmospherePressure" to atmosphere.totalPressure,
        "oceanVolume" to ocean.volume,
        "uvLevel" to uvLevel,
        "cosmicRayFlux" to cosmicRayFlux,
        "magneticField" to magneticFieldStrength,
    )
}
